flakes: No longer compute revCount for local git repository by default

Mic92 opened the pull request:
Computing a revCount is expensive and might not work if the local history is not complete.
This particular fixes a long standing crash bug in Nix: https://github.com/NixOS/nix/issues/6073

<!--

IMPORTANT

Nix is a non-trivial project, so for your contribution to be successful,
it really is important to follow the contributing guidelines:

https://github.com/NixOS/nix/blob/master/CONTRIBUTING.md

Even if you've contributed to open source before, take a moment to read it,
so you understand the process and the expectations.

- what information to include in commit messages
- proper attribution
- volunteering contributions effectively
- how to get help and our review process.

-->

## Motivation

Depends on https://github.com/NixOS/nix/pull/13265



<!-- Briefly explain what the change is about and why it is desirable. -->

## Context

<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

roberth commented on src/libfetchers/git-utils.cc:
```diff
@@ -322,8 +322,11 @@ struct GitRepoImpl : GitRepo, std::enable_shared_from_this<GitRepoImpl>
-                if (git_commit_parent(&parent, commit->get(), n))
-                    throw Error("getting parent of Git commit '%s': %s", *git_commit_id(commit->get()), git_error_last()->message);
+                if (git_commit_parent(&parent, commit->get(), n)) {
+                    // If the parent commit cannot be found (e.g., in a worktree derived from a shallow clone),
+                    // return the number of commits found so far. This matches the behavior of `git rev-list --all`.
+                    return done.size();
```
We can't continue with incorrect information, as that breaks evaluation purity.
The correct thing to do is to fetch the commit history. That can still be a shallow fetch, with `git fetch --filter=tree:none`, i.e. fetching about the minimum amount of objects that is needed for the count.

Alternatively / additionally the impact of this error could be reduced by computing the `revCount` attribute lazily. That would also increase performance in the general case where `revCount` isn't evaluated.

Mic92 replied:
I just checked the history and before we switched to libgit2 we also used `git rev-list --count <commit>`:

```
$ git rev-list --count HEAD
2
```

So this is a regression and I don't see why we should not stick to the semantics of git.

roberth replied:
Git CLI semantics are bad. We don't support impure things like `export-subst` either.
Fortunately this is still an experimental feature, so we have no reason to choose impure behavior.

> potentially expensive on git infrastructure (turning a shallow clone into a full one).

It does not do that. It only fetches the little commit objects, and no trees or blobs.

> nixpkgs ci

Does nixpkgs ci need a value for `revCount`? Otherwise laziness, as proposed, is another way out.

Mic92 replied:
Sorry I don't think what you are asking is possible to implement? Where should this fetch from? There doesn't have to be a remote set or there might be multiple remote to choose from. The remote might require authentication and network that is not available to the current user that runs nix (i.e. something someone might want to do for security reasons in CI systems).

Mic92 replied:
This took 50s on nixpkgs with a gigabit connection and downloaded.

```
% git fetch --filter=tree:0 --depth=99999 origin master
remote: Enumerating objects: 806036, done.
remote: Counting objects: 100% (806036/806036), done.
remote: Compressing objects: 100% (767969/767969), done.
remote: Total 806036 (delta 38067), reused 806031 (delta 38067), pack-reused 0 (from 0)
Receiving objects: 100% (806036/806036), 336.58 MiB | 43.36 MiB/s, done.
Resolving deltas: 100% (38067/38067), done.
From https://github.com/NixOS/nixpkgs
 * branch                  master     -> FETCH_HEAD
noglob hub fetch --filter=tree:0 --depth=99999 origin master  12.38s user 8.62s system 45% cpu 46.003 total
```

For comparison the original shallow clone only took 9s.

Mic92 replied:
Now it no longer changes anything about the revCount but throws a different error message.

Mic92 committed:
flakes: for detected git repository now assume shallow clones by default

Shallow clones are faster to access because we don't have to compute the
revCount, which in sparse checkouts might not even exists.

This is especially useful in combination with lazy trees in mind on large
repository such as nixpkgs.

DavHau commented:
> If we'd just drop revCount, or make it opt-in, I wouldn't particularly mind it.

I fully agree. `revCount` should be remove from flakes and `fetchTree`.
(alternatively opt-in)

As expensive as computing `revCount` is, it doesn't justify the benefit. Its only use case seems to be generating a version string dynamically, but the `shortRev` can be used instead.

grahamc commented:
> Flake regression is unrelated (500er http github error).

On that note, FlakeHub prohibits a hard revcount dependency because it loses that data once it is published. In other words, it is unlikely a primarily-flakehub backed corpus would fail if the feature went away.

edolstra commented:
FWIW, https://github.com/DeterminateSystems/nix-src/pull/245 makes `revCount` calculation a lot faster.

I looked into making the `revCount` computation lazy, but that's not really possible because the evaluation cache fingerprint depends on it.

xokdvium commented:
The lazy revCount approach has been implemented in https://github.com/NixOS/nix/pull/15772.
